src/devices/machine/spi_sdcard.cpp

```diff
@@ -193,163 +216,214 @@ void spi_sdcard_device::spi_clock_w(int state)
+			m_data[10] = 'D';
+			m_data[11] = 0x10; // Product Revision in BCD (1.0)
+			{
+				u32 uSerial = 0x12345678;
+				m_data[12] = (uSerial >> 24) & 0xff; // PSN - Product Serial Number
+				m_data[13] = (uSerial >> 16) & 0xff;
+				m_data[14] = (uSerial >> 8) & 0xff;
+				m_data[15] = (uSerial & 0xff);
+			}
+			m_data[16] = 0x01; // MDT - Manufacturing Date
+			m_data[17] = 0x59; // 0x15 9 = 2021, September
+			m_data[18] = 0x00; // CRC7, bit 0 is always 0
+			{
+				util::crc16_t crc16 = util::crc16_creator::simple(&m_data[3], 16);
+				m_data[19] = (crc16 >> 8) & 0xff;
+				m_data[20] = (crc16 & 0xff);
+			}
+			send_data(3 + 16 + 2, SD_STATE_STBY);
+			break;
 
-	case 16: // CMD16 - SET_BLOCKLEN
-		m_blksize = (m_cmd[3] << 8) | m_cmd[4];
-		if (hard_disk_set_block_size(m_harddisk, m_blksize))
-		{
+		case 12: // CMD12 - STOP_TRANSMISSION
 			m_data[0] = 0;
-		}
-		else
-		{
-			m_data[0] = 0xff;   // indicate an error
-			// if false was returned, it means the hard disk is a CHD file, and we can't resize the
-			// blocks on CHD files.
-			logerror("spi_sdcard: Couldn't change block size to %d, wrong CHD file?", m_blksize);
-		}
-		send_data(1);
-		break;
+			send_data(1,
+					  m_state == SD_STATE_RCV ? SD_STATE_PRG : SD_STATE_TRAN);
+			break;
 
-	case 17: // CMD17 - READ_SINGLE_BLOCK
-		if (m_harddisk)
-		{
-			m_data[0] = 0x00; // initial R1 response
-			// data token occurs some time after the R1 response.  A2SD expects at least 1
-			// byte of space between R1 and the data packet.
-			m_data[2] = 0xfe; // data token
-			u32 blk = (m_cmd[1] << 24) | (m_cmd[2] << 16) | (m_cmd[3] << 8) | m_cmd[4];
-			if (m_type == SD_TYPE_V2)
+		case 16: // CMD16 - SET_BLOCKLEN
+			m_blksize = (m_cmd[3] << 8) | m_cmd[4];
+			if (hard_disk_set_block_size(m_harddisk, m_blksize))
 			{
-				blk /= m_blksize;
+				m_data[0] = 0;
 			}
-			LOGMASKED(LOG_GENERAL, "reading LBA %x\n", blk);
-			hard_disk_read(m_harddisk, blk, &m_data[3]);
+			else
 			{
-				util::crc16_t crc16 = util::crc16_creator::simple(&m_data[3], m_blksize);
-				m_data[m_blksize + 3] = (crc16 >> 8) & 0xff;
-				m_data[m_blksize + 4] = (crc16 & 0xff);
+				m_data[0] = 0xff; // indicate an error
+				// if false was returned, it means the hard disk is a CHD file, and we can't resize the
+				// blocks on CHD files.
+				logerror("spi_sdcard: Couldn't change block size to %d, wrong "
+						 "CHD file?",
+						 m_blksize);
 			}
-			send_data(3 + m_blksize + 2);
-		}
-		else
-		{
-			m_data[0] = 0xff; // show an error
-			send_data(1);
-		}
-		break;
+			send_data(1, SD_STATE_TRAN);
+			break;
+
+		case 17: // CMD17 - READ_SINGLE_BLOCK
+			if (m_harddisk)
+			{
+				m_data[0] = 0x00; // initial R1 response
+				// data token occurs some time after the R1 response.  A2SD expects at least 1
+				// byte of space between R1 and the data packet.
+				m_data[1] = 0xff;
+				m_data[2] = 0xfe; // data token
+				u32 blk = (m_cmd[1] << 24) | (m_cmd[2] << 16) | (m_cmd[3] << 8) | m_cmd[4];
```

Comment:
You should explicitly cast the `u8` values to `u32` before doing the shifts – C automatic promotion rules are annoying, and this will end up causing Coverity warnings.